### PR TITLE
MES-1260 - Added Authentication

### PR DESCRIPTION
**Pull request by Cavalier07**
## Description and relevant Jira numbers

Adds authentication to the app.

## Pull Request checklist:

- [ ] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop
- [ ] commits are squashed

## Sign off process checklist:

- [x] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [x] Reviewers selected in Github
- [x] Any new reusable component/widget added to component library on Confluence

- [ ] Screenshot(s) captured on the physical device (if applicable)
- [ ] Tested by QA
- [x] PO's approval

### CODE REVIEW

File: src/classes/base-page.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { Platform, NavController } from 'ionic-angular';
+import { AuthenticationProvider } from '../providers/authentication/authentication';
+
+export abstract class BasePageComponent {
```
Comment (ammarhaiderbjss) [inline src/classes/base-page.ts]: do we know what this keyword does?

File: src/providers/authentication/authentication.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+  constructor(
+    private msAdal: MSAdal,
+    private inAppBrowser: InAppBrowser,
+    appConfig: AppConfigProvider) {
+    this.authenticationSettings = appConfig.getAppConfig().authentication;
+  }
+
+  isAuthenticated = (): boolean => {
+    return this.authenticationToken ? true : false;
```
Comment (ammarhaiderbjss) [inline src/providers/authentication/authentication.ts]: Need to have a think about how to make this a better check for whether a user is logged in or not, should have a look in the adal library's source code to see if there's anything we can use there. Failing that, we could decode the token and check for some essential values.

File: src/classes/base-page.ts (hunk 1 above)
Comment (Cavalier07) [reply src/classes/base-page.ts]: abstract ? - Basically we are unable to create an instance of it, the only way to use it is to create a sub class (All pages will extend this class). We are basically using it give the same functionality to multiple pages and if needed in the future can use it to force pages to implement certain methods.

File: src/providers/authentication/authentication.ts (hunk 2 above)
Comment (Cavalier07) [reply src/providers/authentication/authentication.ts]: Agreed - This is something we need to investigate - could it be worth spinning this off into a second story - Azure AD Token Validation in Mobile App . I'm wary of pushing more into this  ticket

File: src/pages/login/__tests__/login.spec.ts [hunk 3]
```diff
@@ -34,6 +37,11 @@ describe('LoginPage', () => {
     it('should create', () => {
       expect(component).toBeDefined();
     });
+    it('should login successfully', () => {
+      component.login().then(() => {
+        expect(navController.setRoot).toBeCalled;
```
Comment (johnmorleybjss) [inline src/pages/login/__tests__/login.spec.ts]: The call seems to be missing here, I don't think you're checking anything. Would be nice to test it's called with the correct argument too.

File: src/providers/app-config/app-config.ts [hunk 4]
```diff
@@ -34,6 +34,13 @@ export class AppConfigProvider {
       this.appConfig = {
```
Comment (johnmorleybjss) [inline src/providers/app-config/app-config.ts]: Perhaps we could do something about this config call failing? I imagine it's quite a common scenario (e.g. when offline).

File: src/providers/authentication/__mocks__/ms-adal.mock.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+// TODO: Move it to providers/authentication/__mocks__/
```
Comment (johnmorleybjss) [inline src/providers/authentication/__mocks__/ms-adal.mock.ts]: Think you've addressed this TODO.

File: src/providers/authentication/__tests__/authentication.spec.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import { AuthenticationProvider } from '../authentication';
+import { AppConfigProvider } from '../../app-config/app-config';
+import { AppConfigProviderMock } from '../../app-config/__mocks__/app-config.mock';
+import { InAppBrowserMock } from '../__mocks__/in-app-browser.mock';
+
+describe('Authentication', () => {
+  let authenticationProvider: AuthenticationProvider;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
```
Comment (johnmorleybjss) [inline src/providers/authentication/__tests__/authentication.spec.ts]: Do you need TestBed to test this class? Or is it just easier than managing the mock lifecycle by hand?

File: src/providers/authentication/__tests__/authentication.spec.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    it('should sign in with credetials', async() => {
+      await authenticationProvider.loginWithCredentials();
+
+      expect(authenticationProvider.isAuthenticated()).toEqual(true);
+      expect(authenticationProvider.getAuthenticationToken()).toEqual('AYSNC TEST TOKEN');
+    });
+
+    it('should do something when login fails', async () => {
+      // TODO
```
Comment (johnmorleybjss) [inline src/providers/authentication/__tests__/authentication.spec.ts]: Seems quite an important test to elaborate.

Comment (Cavalier07) [reply src/providers/authentication/__tests__/authentication.spec.ts]: The function in the authentication provide it doesn't do anything either, I guess we can use it for logging or retries if we decide to implement them. It didn't seem right to restrict access/redirect people from here (I think it's better the pages control that) - The other option was to remove that test but I think it works as a reminder this is something we need to look at in the future

File: package.json [hunk 8]
```diff
@@ -43,19 +43,25 @@
...
     "cordova-plugin-device": "^2.0.2",
+    "cordova-plugin-inappbrowser": "3.0.0",
     "cordova-plugin-ionic-keyboard": "^2.1.3",
     "cordova-plugin-ionic-webview": "^2.2.5",
+    "cordova-plugin-ms-adal": "0.10.1",
     "cordova-plugin-splashscreen": "^5.0.2",
     "cordova-plugin-statusbar": "^2.4.2",
     "cordova-plugin-whitelist": "^1.3.3",
     "event-stream": "3.3.4",
+    "io": "1.0.0",
```
[the remark this answers is not in the document]
Comment (Cavalier07) [reply package.json]: My guess is it snuck in have removed it and it didn't seem to break anything

File: src/pages/login/__tests__/login.spec.ts (hunk 3 above)
Comment (Cavalier07) [reply src/pages/login/__tests__/login.spec.ts]: Fixed

**Pull request merged**